cooper/cmp.py

```diff
@@ -1,42 +1,44 @@
 import abc
-from collections.abc import Sequence
-from dataclasses import dataclass
-from typing import Optional
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from typing import Any, Iterator, Literal, Optional
 
 import torch
 
 from cooper.constraints import Constraint, ConstraintMeasurement, ConstraintState
-
-# Formulation, and some other classes below, are inspired by the design of the
-# TensorFlow Constrained Optimization (TFCO) library:
-# https://github.com/google-research/tensorflow_constrained_optimization
+from cooper.multipliers import Multiplier, PenaltyCoefficient
 
 
 @dataclass
 class LagrangianStore:
-    """Stores the value of the (primal) Lagrangian, the dual Lagrangian, as well as
-    ConstraintStores for the primal-contributing and dual-contributing constraints."""
+    """Stores the value of the (primal or dual) Lagrangian, as well as
+    ``ConstraintMeasurement``s for the contributing constraints.
+    """
 
-    lagrangian: torch.Tensor
-    dual_lagrangian: Optional[torch.Tensor] = None
-    primal_constraint_measurements: Optional[list[ConstraintMeasurement]] = None
-    dual_constraint_measurements: Optional[list[ConstraintMeasurement]] = None
+    lagrangian: Optional[torch.Tensor] = None
+    measured_constraints: dict[Constraint, ConstraintMeasurement] = field(default_factory=dict)
 
-    def multiplier_values_for_primal_constraints(self):
-        return [_.multiplier_value for _ in self.primal_constraint_measurements]
+    def backward(self) -> None:
+        """Triggers backward calls to compute the gradient of the Lagrangian with
+        respect to the primal variables."""
+        if self.lagrangian is not None and isinstance(self.lagrangian, torch.Tensor):
```

Comment:
Why would the Lagrangian not be a Tensor?
This check sounds more like an `assert isinstante(...)` to me

Reply:
resolved in #90